spam/spam_tutorial.py

```diff
@@ -140,234 +140,400 @@
+# We'll focus on writing LFs for `HAM`, since we identified `SPAM` comments above.
+#
+# First, we'll introduce Snorkel `Preprocessor`s.
+# Just like a `LabelingFunction` is constructed from a black-box Python function that maps a data point to an integer label,
+# a `Preprocessor` is a black-box Python function that maps a data point to a new data point.
+# `LabelingFunction`s can use `Preprocessor`s, which lets us write LFs over transformed or enhanced data points.
+# We use the `@preprocessor(...)` decorator to add extra functionality to preprocessing functions, such as memoization.
+# We'll start by creating a `Preprocessor` that runs `TextBlob` on our comments, then extracts the polarity and subjectivity scores.
 
 # %%
-from snorkel.analysis.error_analysis import error_buckets
+from snorkel.labeling.preprocess import preprocessor
+from textblob import TextBlob
+
+
+@preprocessor(memoize=True)
+def textblob_sentiment(x):
+    scores = TextBlob(x.text)
+    x.polarity = scores.sentiment.polarity
+    x.subjectivity = scores.sentiment.subjectivity
+    return x
+
+
+# %% [markdown]
+# We can use a preprocessor on its own as well.
+# In order to see how we should use TextBlob scores in an LF, let's see how the distributions differ for `SPAM` and `HAM`.
+# We'll have to tune the output of our LFs based on the TextBlob scores.
+# Tuning input parameters or thresholds from model outputs is a common practice in developing LFs.
+
+# %%
+import matplotlib.pyplot as plt
+
+spam_polarities = [
+    textblob_sentiment(x).polarity for _, x in df_dev.iterrows() if x.label == SPAM
+]
+
+ham_polarities = [
+    textblob_sentiment(x).polarity for _, x in df_dev.iterrows() if x.label == HAM
+]
 
-buckets = error_buckets(golds=Y_dev, preds=L_dev_array)
+plt.hist([spam_polarities, ham_polarities])
+plt.title("TextBlob sentiment polarity scores")
+plt.xlabel("Sentiment polarity score")
+plt.ylabel("Count")
+plt.legend(["Spam", "Ham"])
+plt.show()
 
-df_dev[["text", "label"]].iloc[buckets[(SPAM, HAM)]].head()
 
 # %% [markdown]
-# On the other hand, `buckets[(SPAM, SPAM)]` points to `SPAM` data points that the LF labeled correctly.
+# Using our accuracy-over-coverage principle above, we'll target the high polarity bin on the far right in our LF since there are many more `HAM` comments.
 
 # %%
-df_dev[["text", "label"]].iloc[buckets[(SPAM, SPAM)]].head()
+@labeling_function(preprocessors=[textblob_sentiment])
+def textblob_polarity(x):
+    return HAM if x.polarity > 0.9 else ABSTAIN
+
 
 # %% [markdown]
-# And `buckets[(ABSTAIN, SPAM)]` points to data points that the LF abstained on that are actually `SPAM`.
-# Many of these will be best captured by a separate LF, but browsing these examples can be a good check that your LF is capturing most of the examples that you intended it to.
+# Let's do the same for the subjectivity scores.
+# This will run faster than the last cell, since we memoized the `Preprocessor` outputs.
```

Comment:
add a parenthetical explaining what memoize is in case someone hasn't taken a data structures class?